### PR TITLE
[IndVars] Preserve flags of narrow IV inc if replacing with wider inc.

### LINKED ISSUE #71517: [AArch64] Missed vectorisation opportunity (tsvc, s172)
We are not vectorising kernel s172 from TSVS and are 3x behind compared to GCC as a result. Compile this input with `-O3 -ffast-math -mcpu=neoverse-v2`:

```
__attribute__((aligned(64))) float x[32000];

__attribute__((aligned(64))) float a[32000],b[32000],c[32000],d[32000],e[32000],
                                   aa[256][256],bb[256][256],cc[256][256],tt[256][256];

int dummy(float[32000], float[32000], float[32000], float[32000], float[32000], float[256][256], float[256][256], float[256][256], float);

float  s172(int xa, int xb)
{
    int n1 = xa;
    int n3 = xb;

    for (int nl = 0; nl < 100000; nl++) {
        for (int i = n1-1; i < 32000; i += n3) {
            a[i] += b[i];
        }
        dummy(a, b, c, d, e, aa, bb, cc, 0.);
    }
}
```

Clang's codegen:

```
.LBB0_3:                                //   Parent Loop BB0_2 Depth=1
        ldr     s0, [x19, x8, lsl #2]
        ldr     s1, [x20, x8, lsl #2]
        fadd    s0, s1, s0
        str     s0, [x20, x8, lsl #2]
        add     x8, x8, x22
        cmp     x8, x23
        b.lt    .LBB0_3
```

GCC's codegen:

```
        whilelo p7.s, wzr, w28
.L5:
        ld1w    z31.s, p7/z, [x19, x0, lsl 2]
        ld1w    z30.s, p7/z, [x27, x0, lsl 2]
        fadd    z31.s, z31.s, z30.s
        st1w    z31.s, p7, [x19, x0, lsl 2]
        add     x0, x0, x2
        whilelo p7.s, w0, w28
        b.any   .L5
```

See also: 
https://godbolt.org/z/W8eEPKqET

TODO: root cause analysis.

### PR DESCRIPTION
**Pull request by fhahn**
We are replacing a narrow IV increment with a wider one. If the original (narrow) increment did not wrap, the wider one should not wrap either. Set the flags to be the union of both wide increment and original increment; this ensures we preserve flags SCEV could infer for the wider increment.

Fixes https://github.com/llvm/llvm-project/issues/71517.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
(This question also goes for the similar change at https://github.com/llvm/llvm-project/pull/79512...)

Are we guaranteed matching increments at this point? I.e. do we know that both instructions actually increment by the same amount, and we don't have, say, a `iv.next = iv - 7 + 12` sequence on one end and `iv.next = iv + 5` on the other?

**Comment by fhahn** [comment]
> (This question also goes for the similar change at #79512...)
> 
> Are we guaranteed matching increments at this point? I.e. do we know that both instructions actually increment by the same amount, and we don't have, say, a `iv.next = iv - 7 + 12` sequence on one end and `iv.next = iv + 5` on the other?

I couldn't find a case where this would happen and cause incorrect use of flags by adding various assertions, but in theory I think it may happen. Updated the code to check that the increments both directly operate on the respective phis and have the same opcode. this should also imply the same step (module extensions) assuming the start is the same.

**Comment by fhahn** [comment]
@nikic WDYT of the updated checks? Not super happy with how they look, any ideas/suggestions how to improve them? Once we converged, I'll something similar as follow-up to https://github.com/llvm/llvm-project/pull/79512.

File: llvm/lib/Transforms/Utils/SimplifyIndVar.cpp [hunk 1]
```diff
@@ -1985,7 +1985,28 @@ PHINode *WidenIV::createWideIV(SCEVExpander &Rewriter) {
       // increment to the new (widened) increment.
       auto *OrigInc =
           cast<Instruction>(OrigPhi->getIncomingValueForBlock(LatchBlock));
+
       WideInc->setDebugLoc(OrigInc->getDebugLoc());
+      // We are replacing a narrow IV increment with a wider IV increment . If
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:1990]:
```suggestion
      // We are replacing a narrow IV increment with a wider IV increment. If
```

**Review by nikic — APPROVED** [review]
LGTM, thanks!

File: llvm/lib/Transforms/Utils/SimplifyIndVar.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Utils/SimplifyIndVar.cpp]: Thanks, should be fixed!

**Pull request merged**